content/blog/2020-12-18-futil-passes/index.md

```diff
@@ -0,0 +1,146 @@
+- **a `cells` section**, containing the hardware subcomponents contained in the program. Those subcomponents are hereafter in this post called `cell`s. For example, the FuTIL standard library contains definitions for `cell`s such as memories or registers.
+- **a `wires` section**, specifying the wires that connect `cell`s to one another via cell _ports_. For example, a `wires` section might contain the statement `x.in = y.out;`, which would represent connecting a wire between the `out` port of `y`, and the `in` port of `x`. These definitions can be organized into `group`s.
+- **a `control` section**, defining when `group`s execute. FuTIL supports various control constructs, like `while` loops and `if`/`else` statements, as well as `par` and `seq` statements, to signal to execute `group`s in parallel or in sequence.
+
+
+## Static Timing Inference Compiler Pass
+
+### Introduction
+
+The Calyx compiler features an optimization pass that leverages static knowledge about `group` latencies. Generated programs are sped up, because the compiler doesn't need to insert extra logic to wait for a `group` to finish, if it already knows the latency of the `group`. Frontends that are targeting FuTIL can supply this information with `group` annotations:
+
+    group a<"static"=5> {
+      ...
+    }
+
+The static timing inference pass attempts to infer this `"static"` value and insert it automatically, or error out if its inferred value doesn't match an already existing (and then necessarily incorrect) annotation.
+
+### Inferring `group` Latencies
+
+FuTIL `cell`s have _go/done_ interfaces. The `go` port for a `cell`, when _high_, indicates that the `cell` should start its computation. When the `cell` finishes executing, it will make its `done` port high. The names of these ports can be anything. For example, the FuTIL standard library contains a [definition for a register `cell`][registers], which has a `go` port called `write_en`, indicating that it's time to write a value, and a `done` port, indicating that the write is complete.
+
+For some `cell` definitions, we know statically the number of cycles it takes from the moment a `go` signal is raised to when the `done` signal is raised. For registers, this latency is a single cycle. Starting with this knowledge, it's possible to deduce the latencies of some `group`s. This pass targets `group`s where these conditions are true:
+- All `go` ports are written to only by `done` ports or constants.
+- All `done` ports are only written to by other `done` signals.
+- The latency of each cell in this chain of `go`/`done` signals is known statically.
+
+For example, this group, called `a`, takes only a single cycle to execute:
+
+    group a {
+      register1.write_en = 1'd1;
+      a[done] = register1.done;
+    }
+
+The `go` port of `register1` is driven by a constant, and the `done` signal of the group is written by the `done` signal of `register1`.
```

Comment:
Very clear description, but maybe it would be useful to explain why you're making these restrictions (i.e., handing the general case is probably uncomputable, so you need to pick some sort of restriction, and this seems to cover plenty of cases).

Reply:
Added some elaboration on the reason for the restrictions!